zashi-pipeline.py

```diff
@@ -0,0 +1,217 @@
+
+    # Fetch the issues within the graph.
+    mapping = github.download_issues(gapi, dg.nodes, repos)
+
+    # Relabel the graph
+    dg = nx.relabel_nodes(dg, mapping)
+
+    # Filter out unknown issues
+    unknown = [n for n in dg if n.repo_id not in repos]
+    if len(unknown) > 0:
+        dg.remove_nodes_from(unknown)
+
+    # Apply property annotations
+    for (source, sink) in dg.edges:
+        attrs = dg.edges[source, sink]
+        attrs['is_open'] = 0 if source.state == 'closed' else 1
+
+    # Render the HTML version!
+    html_header = '''<!DOCTYPE html>
+<html>
+  <head>
+    <title>Zashi Pipeline</title>
+
+    <style>
+      body {
+        color: #1f2328;
+        font-family: -apple-system,BlinkMacSystemFont,"Segoe UI","Noto Sans",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji";
+        font-size: 14px;
+        line-height: 1.5;
+        word-wrap: break-word;
+      }
+      a {
+        color: #0969da;
+      }
+      table {
+        border-collapse: collapse;
+        width: 100%;
+        width: max-content;
+        max-width: 100%;
+        overflow: auto;
+      }
+      table tr {
+        border-top: 1px solid #d1d9e0b3;
+      }
+      table th, table td {
+        border: 1px solid #d1d9e0b3;
+        padding: 6px 13px;
+        text-align: center;
+      }
+
+      @media (prefers-color-scheme: dark) {
+        body {
+          background-color: #121212;
+          color: #f0f6fc;
+        }
+        a {
+          color: #4493f8;
+        }
+        table tr {
+            border-top: 1px solid #3d444db3;
+        }
+        table th, table td {
+            border: 1px solid #3d444db3;
+        }
+      }
+    </style>
+  </head>
+  <body>
+    <h1>Zashi Pipeline</h1>
+    <p>🐞 = bug, 💡 = feature, ✅ = implemented / released, 🛑 = unfinished, 📥 = unassigned / DAG needs updating</p>
+    <table>
+      <thead>
+        <tr>
+          <th>Type</th>
+          <th>Issue</th>
+          <th>Rust crate</th>
+          <th>Android SDK</th>
+          <th>Swift SDK</th>
+          <th>Zashi Android</th>
+          <th>Zashi iOS</th>
+        </tr>
+      </thead>
+      <tbody>
+'''
+    html_footer = '''
+      </tbody>
+    </table>
+  </body>
+</html>
+'''
+    with open('public/zashi-pipeline.html', 'w') as f:
+        f.write(html_header)
+
+        for issue in tracked_issues.values():
+            f.write('<tr>')
+
+            if 'C-tracked-bug' in issue.labels:
+                f.write('<td>🐞</td>')
+            else:
+                f.write('<td>💡</td>')
+
+            f.write('<td>{} <a href="{}">{}</a></td>'.format(
+                '✅' if issue.state == 'closed' else '🛑',
+                issue.url,
+                issue.title,
+            ))
+
+            children = nx.descendants(dg, issue)
+            def find_child_release(repo_id):
+                for child in children:
+                    if child.repo_id == repo_id and 'C-release' in child.labels:
```

Comment:
Current limitation: this doesn't reliably find the correct Rust crate release when we have multiple release issues created in advance. We need to instead:
- Find every descendant that potentially matches.
- Find the shortest path from the tracked issue to each descendant.
- Render the descendant(s) with the shortest path length (which for a well-formed DAG should always be at most one breaking change release, and potentially also a backwards-compatible release in the case of hotfixes).

Reply:
Is it possible to do this as a depth-first search instead, finding the first `C-release` issue along each branch? That seems a bit simpler, and would also allow us to handle the case where an issue is blocking both a hotfix and a semver-breaking release.